acoular/spectra.py

```diff
@@ -392,13 +409,23 @@ def calc_csm(self):
             else:
                 raise ValueError('Calibration data not compatible: %i, %i' % (self.calib.num_mics, t.numchannels))
         # get time data blockwise
-        for data in self.get_source_data():
-            ft = fft.rfft(data * wind, None, 0).astype(self.precision)
-            calcCSM(csm_upper, ft)  # only upper triangular part of matrix is calculated (for speed reasons)
-        # create the full csm matrix via transposing and complex conj.
-        csm_lower = csm_upper.conj().transpose(0, 2, 1)
-        [fill_diagonal(csm_lower[cntFreq, :, :], 0) for cntFreq in range(csm_lower.shape[0])]
-        csm = csm_lower + csm_upper
+        full_blocks = int(ceil(self.num_blocks))
+        n_groups = int(ceil(self.source.numsamples / (full_blocks * self.block_size)))
+        csm = zeros(shape=csm_shape)
+        csm = csm[:, newaxis, :, :]
+        for _ in range(n_groups):
+            for __, data in zip(range(full_blocks), self.get_source_data()):
+                ft = fft.rfft(data * wind, None, 0).astype(self.precision)
+                calcCSM(csm_upper, ft)  # only upper triangular part of matrix is calculated (for speed reasons)
+            # create the full csm matrix via transposing and complex conj.
+            csm_lower = csm_upper.conj().transpose(0, 2, 1)
+            [fill_diagonal(csm_lower[cntFreq, :, :], 0) for cntFreq in range(csm_lower.shape[0])]
+            csm_block = csm_lower + csm_upper
+            if n_groups > 1:
+                csm_block = csm_block[:, newaxis, :, :]
+                csm = concatenate([csm, csm_block], axis=1)
+            else:
+                csm = csm_block
```

Comment:
Two problems here:

1. breaks API by possibly returning a four-dimensional array. What we need is just one CSM per frequency, and the num_blocks just helps to limit the number of blocks (one time) in case we have an infinite stream as the input. Full streaming cannot be done anyway with the `BeamformerBase` derived classes.
2. the code is very tortuous and difficult to check
```suggestion
        for num, data in enumerate(self.get_source_data()):
            if num > self.num_blocks:
                break
            ft = fft.rfft(data * wind, None, 0).astype(self.precision)
            calcCSM(csm_upper, ft)  # only upper triangular part of matrix is calculated (for speed reasons)
        # create the full csm matrix via transposing and complex conj.
        csm_lower = csm_upper.conj().transpose(0, 2, 1)
        [fill_diagonal(csm_lower[cntFreq, :, :], 0) for cntFreq in range(csm_lower.shape[0])]
        csm = csm_lower + csm_upper
        # in case there where less blocks available
        if num < self.num_blocks:
            return csm * (2.0 / self.block_size / weight / num)
```

Reply:
Reverted: [8d41784](https://github.com/acoular/acoular/pull/285/commits/8d417847fb9a19be7377bb2cc2756182359a5fdd)

On a side-note: I thought that an extra dimension was specifically requested by @adku1173, but that must have been a misunderstanding. According to 

> There is no need to adjust the number of blocks that the `PowerSpectra` object should average. Instead, the classes `MaskedTimeSamples` and `MaskedTimeInOut` exists, which can be used to set the evaluation time interval. Thus, I suggest to revert this change.

this is redundant anyway and changes are being made on a lower level: #294